File: src/ios/Founditure/Services/AuthService.swift
// MARK: - Human Tasks
/*
 TODO: Human Configuration Required
 1. Configure JWT token refresh thresholds in production environment
 2. Set up proper keychain access groups for token storage
 3. Configure biometric authentication if required
 4. Set up proper SSL certificate pinning
 5. Configure proper token expiration monitoring
*/

import Foundation // iOS 14.0+
import Combine // iOS 14.0+

// MARK: - Auth Response Model
/// Model representing authentication response from server
struct AuthResponse: Codable {
    let accessToken: String
    let refreshToken: String
    let expiresIn: Int
    let userId: String
}

// MARK: - Auth Service Protocol
/// Protocol defining authentication service interface
@available(iOS 14.0, *)
protocol AuthServiceProtocol {
    func login(email: String, password: String) -> AnyPublisher<AuthResponse, Error>
    func register(email: String, password: String, fullName: String) -> AnyPublisher<AuthResponse, Error>
    func logout() -> AnyPublisher<Void, Error>
}

// MARK: - Auth Service Implementation
/// Service handling all authentication-related operations
/// Requirement: User Authentication - Implements secure user authentication and authorization for iOS client
@available(iOS 14.0, *)
final class AuthService: ServiceProtocol, AuthServiceProtocol {
    // MARK: - Properties
    private let apiClient: APIClient
    private let baseURL: String
    private let session: URLSession
    private var refreshTimer: Timer?
    private let tokenSubject = CurrentValueSubject<String?, Never>(nil)
    
    // Constants for token storage
    private enum TokenKeys {
        static let accessToken = "auth.accessToken"
        static let refreshToken = "auth.refreshToken"
    }
    
    // MARK: - Initialization
    /// Initializes the auth service with required dependencies
    /// Requirement: Security Architecture - Initializes secure authentication components
    init(apiClient: APIClient, baseURL: String) {
        self.apiClient = apiClient
        self.baseURL = baseURL
        
        // Configure secure session
        let configuration = URLSessionConfiguration.default
        configuration.tlsMinimumSupportedProtocol = .tlsProtocol12
        self.session = URLSession(configuration: configuration)
        
        // Attempt to restore existing session
        restoreSession()
    }
    
    // MARK: - Authentication Methods
    /// Authenticates user with email and password
    /// Requirement: Authentication Flow - Implements secure login flow
    func login(email: String, password: String) -> AnyPublisher<AuthResponse, Error> {
        // Validate credentials
        let validationResult = ValidationHelper.validateUserCredentials(email: email, password: password)
        
        switch validationResult {
        case .success:
            // Create login request
            return apiClient.request(
                endpoint: "auth/login",
                method: .post,
                body: ["email": email, "password": password]
            )
            .flatMap { [weak self] (response: AuthResponse) -> AnyPublisher<AuthResponse, Error> in
                guard let self = self else {
                    return Fail(error: ServiceError.unknownError).eraseToAnyPublisher()
                }
                
                // Store tokens securely
                return self.storeTokens(response)
                    .map { response }
                    .eraseToAnyPublisher()
            }
            .eraseToAnyPublisher()
            
        case .failure(let error):
            return Fail(error: error).eraseToAnyPublisher()
        }
    }
    
    /// Registers new user account
    /// Requirement: User Authentication - Implements secure user registration
    func register(email: String, password: String, fullName: String) -> AnyPublisher<AuthResponse, Error> {
        // Validate registration data
        let validationResult = ValidationHelper.validateUserCredentials(email: email, password: password)
        
        switch validationResult {
        case .success:
            // Create registration request
            return apiClient.request(
                endpoint: "auth/register",
                method: .post,
                body: [
                    "email": email,
                    "password": password,
                    "fullName": fullName
                ]
            )
            .flatMap { [weak self] (response: AuthResponse) -> AnyPublisher<AuthResponse, Error> in
                guard let self = self else {
                    return Fail(error: ServiceError.unknownError).eraseToAnyPublisher()
                }
                
                // Store tokens securely
                return self.storeTokens(response)
                    .map { response }
                    .eraseToAnyPublisher()
            }
            .eraseToAnyPublisher()
            
        case .failure(let error):
            return Fail(error: error).eraseToAnyPublisher()
        }
    }
    
    /// Logs out current user and clears session
    /// Requirement: Authentication Flow - Implements secure logout
    func logout() -> AnyPublisher<Void, Error> {
        // Stop refresh timer
        refreshTimer?.invalidate()
        refreshTimer = nil
        
        // Clear stored tokens
        return clearTokens()
            .handleEvents(receiveCompletion: { [weak self] completion in
                if case .finished = completion {
                    self?.tokenSubject.send(nil)
                }
            })
            .eraseToAnyPublisher()
    }
    
    // MARK: - Token Management
    /// Refreshes authentication token before expiration
    /// Requirement: Security Architecture - Implements JWT service and token refresh
    private func refreshToken() -> AnyPublisher<AuthResponse, Error> {
        // Retrieve refresh token
        let refreshResult = KeychainManager.retrieve(key: TokenKeys.refreshToken)
        
        switch refreshResult {
        case .success(let tokenData):
            guard let refreshToken = String(data: tokenData, encoding: .utf8) else {
                return Fail(error: ServiceError.invalidResponse).eraseToAnyPublisher()
            }
            
            // Create refresh request
            return apiClient.request(
                endpoint: "auth/refresh",
                method: .post,
                body: ["refreshToken": refreshToken]
            )
            .flatMap { [weak self] (response: AuthResponse) -> AnyPublisher<AuthResponse, Error> in
                guard let self = self else {
                    return Fail(error: ServiceError.unknownError).eraseToAnyPublisher()
                }
                
                // Update stored tokens
                return self.storeTokens(response)
                    .map { response }
                    .eraseToAnyPublisher()
            }
            .eraseToAnyPublisher()
            
        case .failure(let error):
            return Fail(error: error).eraseToAnyPublisher()
        }
    }
    
    /// Securely stores authentication tokens
    /// Requirement: Security Architecture - Implements secure token storage
    private func storeTokens(_ response: AuthResponse) -> AnyPublisher<Void, Error> {
        return Future<Void, Error> { [weak self] promise in
            guard let self = self else {
                promise(.failure(ServiceError.unknownError))
                return
            }
            
            // Store access token
            let accessTokenResult = KeychainManager.save(
                data: Data(response.accessToken.utf8),
                key: TokenKeys.accessToken
            )
            
            // Store refresh token
            let refreshTokenResult = KeychainManager.save(
                data: Data(response.refreshToken.utf8),
                key: TokenKeys.refreshToken
            )
            
            // Check results
            switch (accessTokenResult, refreshTokenResult) {
            case (.success, .success):
                // Update token subject
                self.tokenSubject.send(response.accessToken)
                
                // Configure refresh timer
                self.configureRefreshTimer(expiresIn: response.expiresIn)
                
                promise(.success(()))
                
            case (.failure(let error), _), (_, .failure(let error)):
                promise(.failure(error))
            }
        }.eraseToAnyPublisher()
    }
    
    /// Clears stored authentication tokens
    /// Requirement: Security Architecture - Implements secure token cleanup
    private func clearTokens() -> AnyPublisher<Void, Error> {
        return Future<Void, Error> { promise in
            // Delete access token
            let accessTokenResult = KeychainManager.delete(key: TokenKeys.accessToken)
            
            // Delete refresh token
            let refreshTokenResult = KeychainManager.delete(key: TokenKeys.refreshToken)
            
            // Check results
            switch (accessTokenResult, refreshTokenResult) {
            case (.success, .success):
                promise(.success(()))
            case (.failure(let error), _), (_, .failure(let error)):
                promise(.failure(error))
            }
        }.eraseToAnyPublisher()
    }
    
    /// Configures token refresh timer
    /// Requirement: Authentication Flow - Manages token refresh timing
    private func configureRefreshTimer(expiresIn: Int) {
        refreshTimer?.invalidate()
        
        // Schedule refresh 5 minutes before expiration
        let refreshInterval = TimeInterval(max(expiresIn - 300, 0))
        refreshTimer = Timer.scheduledTimer(withTimeInterval: refreshInterval, repeats: false) { [weak self] _ in
            self?.refreshToken()
                .sink(
                    receiveCompletion: { _ in },
                    receiveValue: { _ in }
                )
                .store(in: &self?.cancellables ?? Set())
        }
    }
    
    /// Restores existing session if available
    /// Requirement: Authentication Flow - Implements session restoration
    private func restoreSession() {
        let accessTokenResult = KeychainManager.retrieve(key: TokenKeys.accessToken)
        
        switch accessTokenResult {
        case .success(let tokenData):
            if let token = String(data: tokenData, encoding: .utf8) {
                tokenSubject.send(token)
                
                // Refresh token to ensure validity
                refreshToken()
                    .sink(
                        receiveCompletion: { _ in },
                        receiveValue: { _ in }
                    )
                    .store(in: &cancellables)
            }
        case .failure:
            tokenSubject.send(nil)
        }
    }
    
    // MARK: - Properties
    private var cancellables = Set<AnyCancellable>()
}

File: src/android/app/src/main/kotlin/com/founditure/android/data/remote/api/AuthService.kt
/*
 * Human Tasks:
 * 1. Ensure Retrofit dependencies are added to app-level build.gradle:
 *    implementation 'com.squareup.retrofit2:retrofit:2.9.0'
 *    implementation 'com.squareup.retrofit2:converter-gson:2.9.0'
 *    implementation 'com.squareup.retrofit2:adapter-rxjava3:2.9.0'
 * 2. Ensure RxJava3 dependencies are added:
 *    implementation 'io.reactivex.rxjava3:rxjava:3.0.0'
 *    implementation 'io.reactivex.rxjava3:rxandroid:3.0.0'
 */

package com.founditure.android.data.remote.api

import com.founditure.android.data.remote.dto.UserDto
import io.reactivex.rxjava3.core.Completable // version: 3.0.0
import io.reactivex.rxjava3.core.Single // version: 3.0.0
import retrofit2.http.Body
import retrofit2.http.POST // version: 2.9.0

/**
 * Retrofit service interface for authentication API endpoints.
 * 
 * Addresses requirements:
 * - Core authentication functionality (1.2 Scope/Core System Components/Backend Services)
 * - Offline-first architecture with auth state management (1.2 Scope/Mobile Applications)
 * - Secure authentication flows (7.1 Authentication and Authorization)
 */
interface AuthService {

    /**
     * Authenticates user with email and password.
     * Returns an observable authentication response containing user data and tokens.
     */
    @POST("auth/login")
    fun login(@Body request: LoginRequest): Single<AuthResponse>

    /**
     * Registers a new user account.
     * Returns an observable authentication response for the new user.
     */
    @POST("auth/register")
    fun register(@Body request: RegisterRequest): Single<AuthResponse>

    /**
     * Refreshes authentication token using refresh token.
     * Returns an observable response with new tokens.
     */
    @POST("auth/refresh")
    fun refreshToken(@Body request: RefreshTokenRequest): Single<AuthResponse>

    /**
     * Invalidates current authentication tokens.
     * Returns an observable completion of logout.
     */
    @POST("auth/logout")
    fun logout(): Completable

    /**
     * Initiates password reset process.
     * Returns an observable completion of password reset request.
     */
    @POST("auth/forgot-password")
    fun forgotPassword(@Body request: ForgotPasswordRequest): Completable

    /**
     * Resets password using reset token.
     * Returns an observable completion of password reset.
     */
    @POST("auth/reset-password")
    fun resetPassword(@Body request: ResetPasswordRequest): Completable
}

/**
 * Request body for login endpoint
 */
data class LoginRequest(
    val email: String,
    val password: String,
    val rememberMe: Boolean
)

/**
 * Request body for registration endpoint
 */
data class RegisterRequest(
    val email: String,
    val password: String,
    val fullName: String,
    val phoneNumber: String?
)

/**
 * Response data for authentication operations
 */
data class AuthResponse(
    val user: UserDto,
    val accessToken: String,
    val refreshToken: String,
    val expiresIn: Long
)

/**
 * Request body for token refresh endpoint
 */
data class RefreshTokenRequest(
    val refreshToken: String
)

/**
 * Request body for forgot password endpoint
 */
data class ForgotPasswordRequest(
    val email: String
)

/**
 * Request body for password reset endpoint
 */
data class ResetPasswordRequest(
    val token: String,
    val newPassword: String
)